lumigator/python/mzai/backend/backend/services/jobs.py

```diff
@@ -173,54 +174,56 @@ def _get_job_params(self, job_type: str, record, request: BaseModel) -> dict:
         return job_params
 
     def _add_dataset_to_db(self, job_id: UUID, request: JobInferenceCreate):
-        loguru.logger.info("Adding a new dataset entry to the database...")
-        s3 = S3FileSystem()
-
-        # Get the dataset from the S3 bucket
-        result_key = self._get_results_s3_key(job_id)
-        with s3.open(f"{settings.S3_BUCKET}/{result_key}", "r") as f:
-            results = json.loads(f.read())
-
-        dataset = {k: v for k, v in results.items() if k in ["examples", request.output_field]}
-
-        # Create a CSV in memory
-        csv_buffer = StringIO()
-        csv_writer = csv.writer(csv_buffer)
-        csv_writer.writerow(dataset.keys())
-        csv_writer.writerows(zip(*dataset.values()))
-
-        # Create a binary file from the CSV, since the upload function expects a binary file
-        bin_data = BytesIO(csv_buffer.getvalue().encode("utf-8"))
-        bin_data_size = len(bin_data.getvalue())
-
-        # Figure out the dataset filename
-        dataset_filename = self.data_service.get_dataset(dataset_id=request.dataset).filename
-        dataset_filename = Path(dataset_filename).stem
-        dataset_filename = f"{dataset_filename}-annotated.csv"
-
-        upload_file = UploadFile(
-            file=bin_data,
-            size=bin_data_size,
-            filename=dataset_filename,
-            headers={"content-type": "text/csv"},
-        )
-        dataset_record = self.data_service.upload_dataset(
-            upload_file,
-            format=DatasetFormat.JOB,
-            run_id=job_id,
-            generated=True,
-            generated_by=results["model"],
-        )
-
-        loguru.logger.info(
-            f"Dataset '{dataset_filename}' with ID '{dataset_record.id}' added to the database."
-        )
+        try:
+            s3 = S3FileSystem()
+
+            # Get the dataset from the S3 bucket
+            result_key = self._get_results_s3_key(job_id)
+            with s3.open(f"{settings.S3_BUCKET}/{result_key}", "r") as f:
+                results = json.loads(f.read())
+
+            dataset = {k: v for k, v in results.items() if k in ["examples", request.output_field]}
+
+            # Create a CSV in memory
```

Comment:
Not sure I understand this comment, we either create a file artifact or work with the code in-memory

Reply:
I didn't write this comment or this code... But I understand that we use a StringIO as a "file-like object" to provide a way to transform a file using the csv package, so it's a file to python code, that is backed by a byte array in memory (and not a data structure). Maybe @aittalam has more info???

Reply:
This is from https://github.com/mozilla-ai/lumigator/pull/522, the main idea is the following:

- we have inference results which are saved by the job as a json file (actual data + metadata)
- we need to store the data part (in particular samples and predictions - I added GT in my follow-up PR) as a dataset that will be ingested by evaluation
- at the moment, only CSV files can be brought on the platform using `DataService.upload_dataset()`, so the data we need is converted into an in-memory CSV and then uploaded

Reply:
Ah sorry I missed that this was not new, thanks for the explanation! So the actual StringIO object is a Python string object that can be passed into csv.writer, which takes any Python object that has a File API, makes sense.